In [76]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test)=imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [78]:
# 텍스트데이터 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad=pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad=pad_sequences(X_test,maxlen=80, padding='post', truncating='post')

In [11]:
from tensorflow.keras import Sequential, layers
model_dnn=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=10000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    layers.Flatten(),
    layers.Dense(64, activation='relu'), #활성화함수 Relu->비선형성
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])


In [12]:
model_dnn.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)         │ (None, 80, 32)         │       320,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 2560)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 64)             │       163,904 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 2)              │           130 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 484,034 (1.85 MB)

 Trainable params: 484,034 (1.85 MB)

 Non-trainable params: 0 (0.00 B)

In [13]:
model_dnn.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_dnn.fit(X_train_pad, y_train, epochs=10, batch_size=200)

Epoch 1/10


2025-09-04 15:23:00.904047: I external/local_xla/xla/service/service.cc:163] XLA service 0x7eb9e80092c0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2025-09-04 15:23:00.904074: I external/local_xla/xla/service/service.cc:171]   StreamExecutor device (0): NVIDIA GeForce RTX 4060 Laptop GPU, Compute Capability 8.9
2025-09-04 15:23:00.941496: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:269] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2025-09-04 15:23:01.055408: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:473] Loaded cuDNN version 91200
2025-09-04 15:23:01.746222: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_117', 124 bytes spill stores, 124 bytes spill loads

2025-09-04 15:23:01.795117: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc

 71/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5451 - loss: 0.6834

I0000 00:00:1756966982.965237   41842 device_compiler.h:196] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.6866 - loss: 0.5670
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8836 - loss: 0.2865
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9782 - loss: 0.0897
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9984 - loss: 0.0163
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9997 - loss: 0.0045
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 0.0019
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 0.0012
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 1.0000 - loss: 8.2270e-04
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 6.0883e-04
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 1.0000 - loss: 4.7041e-04


In [14]:
model_dnn.evaluate(X_test_pad, y_test)

782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7744 - loss: 0.9200


[0.920016348361969, 0.774399995803833]

In [ ]:
# RNN
from tensorflow.keras import Sequential, layers
model_rnn=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=10000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    #layers.Flatten(), 순서 정보가 중요해서 하면안됨
    #layers.Dense(64, activation='relu'), #활성화함수 Relu->비선형성
    layers.SimpleRNN(64),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])

In [20]:
model_rnn.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_4 (Embedding)         │ (None, 80, 32)         │       320,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_1 (SimpleRNN)        │ (None, 64)             │         6,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 2)              │           130 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 326,338 (1.24 MB)

 Trainable params: 326,338 (1.24 MB)

 Non-trainable params: 0 (0.00 B)

In [26]:
model_rnn.compile(optimizer='adam',loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 24ms/step - accuracy: 0.9971 - loss: 0.0111
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 0.9958 - loss: 0.0127
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 0.9948 - loss: 0.0172
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 0.9938 - loss: 0.0205
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 0.9964 - loss: 0.0126
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 22ms/step - accuracy: 0.9988 - loss: 0.0043
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 22ms/step - accuracy: 0.9982 - loss: 0.0069
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 1.0000 - loss: 5.9904e-04
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 22ms/step - accuracy: 1.0000 - loss: 1.6853e-04
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 22ms/step - accuracy: 1.0000 - loss: 8.1100e-05
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7501 - loss: 1.5472


[1.5471824407577515, 0.7500799894332886]

In [24]:
# RNN
from tensorflow.keras import Sequential, layers
model_rnn2=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=10000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    #layers.Flatten(), 순서 정보가 중요해서 하면안됨
    #layers.Dense(64, activation='relu'), #활성화함수 Relu->비선형성
    layers.SimpleRNN(64,return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])

In [ ]:
model_rnn2.compile(optimizer='adam',loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn2.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn2.evaluate(X_test_pad, y_test)

Epoch 1/10


2025-09-04 15:31:53.317616: I external/local_xla/xla/service/gpu/autotuning/dot_search_space.cc:208] All configs were filtered out because none of them sufficiently match the hints. Maybe the hints set does not contain a good representative set of valid configs? Working around this by using the full hints set instead.
2025-09-04 15:31:53.317685: I external/local_xla/xla/service/gpu/autotuning/dot_search_space.cc:208] All configs were filtered out because none of them sufficiently match the hints. Maybe the hints set does not contain a good representative set of valid configs? Working around this by using the full hints set instead.
2025-09-04 15:31:53.317711: I external/local_xla/xla/service/gpu/autotuning/dot_search_space.cc:208] All configs were filtered out because none of them sufficiently match the hints. Maybe the hints set does not contain a good representative set of valid configs? Working around this by using the full hints set instead.
2025-09-04 15:31:53.786933: I external/l

125/125 ━━━━━━━━━━━━━━━━━━━━ 8s 29ms/step - accuracy: 0.5047 - loss: 0.7015
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.6222 - loss: 0.6477
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.8131 - loss: 0.4057
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.9537 - loss: 0.1260
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 0.9889 - loss: 0.0340
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 0.9960 - loss: 0.0130
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.9982 - loss: 0.0055
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.9971 - loss: 0.0090
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.9902 - loss: 0.0271
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.9948 - loss: 0.0156


2025-09-04 15:32:32.972290: I external/local_xla/xla/service/gpu/autotuning/dot_search_space.cc:208] All configs were filtered out because none of them sufficiently match the hints. Maybe the hints set does not contain a good representative set of valid configs? Working around this by using the full hints set instead.
2025-09-04 15:32:33.484956: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_26', 36 bytes spill stores, 36 bytes spill loads



782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.6228 - loss: 2.0055


[2.0054759979248047, 0.6228399872779846]

- RNN: [1.5471824407577515, 0.7500799894332886]
- RNN2: [2.0054759979248047, 0.6228399872779846]

In [27]:
model_rnn2.compile(optimizer='sgd',loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn2.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn2.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 27ms/step - accuracy: 0.9927 - loss: 0.0228
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 0.9992 - loss: 0.0044
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 0.9997 - loss: 0.0024
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 0.9974 - loss: 0.0090
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 0.9987 - loss: 0.0049
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 1.0000 - loss: 0.0013
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 0.9999 - loss: 0.0012
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 0.9998 - loss: 0.0011
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 19ms/step - accuracy: 1.0000 - loss: 7.9282e-04
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 28ms/step - accuracy: 1.0000 - loss: 6.9412e-04
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.6155 - loss: 2.2694


[2.2694077491760254, 0.6155200004577637]

In [28]:
model_rnn2.compile(optimizer='sgd',loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn2.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn2.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 6s 32ms/step - accuracy: 0.9998 - loss: 0.0010    
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 1.0000 - loss: 6.2020e-04
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 0.9988 - loss: 0.0044
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 20ms/step - accuracy: 1.0000 - loss: 7.3754e-04
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 28ms/step - accuracy: 1.0000 - loss: 5.6103e-04
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 28ms/step - accuracy: 1.0000 - loss: 5.2274e-04
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 1.0000 - loss: 4.6793e-04
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 1.0000 - loss: 4.4615e-04
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 1.0000 - loss: 4.6044e-04
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - accuracy: 1.0000 - loss: 4.0360e-04
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.6166 - loss: 2.3756


[2.3755927085876465, 0.6166399717330933]

### 옵티마이저 sgd로 변경

In [79]:
# RNN
from tensorflow.keras import Sequential, layers

model_rnn3=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=10000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    layers.SimpleRNN(64),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])


In [80]:
model_rnn3.compile(optimizer='sgd',loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn3.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn3.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 21ms/step - accuracy: 0.4964 - loss: 0.6973
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.5056 - loss: 0.6940
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.5096 - loss: 0.6935
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.5071 - loss: 0.6933
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - accuracy: 0.5106 - loss: 0.6929
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - accuracy: 0.5112 - loss: 0.6926
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - accuracy: 0.5146 - loss: 0.6924
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.5148 - loss: 0.6922
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.5150 - loss: 0.6920
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - accuracy: 0.5192 - loss: 0.6917
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.5080 - loss: 0.6930


[0.693044126033783, 0.5079600214958191]

### 전체 단어 개수 1000개로 변경

In [81]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test)=imdb.load_data(num_words=1000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 2, 2, 65, 458, 2, 66, 2, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 2, 2, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2, 19, 14, 22, 4, 2, 2, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 2, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2, 2, 16, 480, 66, 2, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 2, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 2, 15, 256, 4, 2, 7, 2, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 2, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2, 56, 26, 141, 6, 194, 2, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 2, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 2, 88, 12, 16, 283, 5, 16, 2, 113, 103, 32, 15, 16, 2, 19, 178, 32]),
         list([1, 194, 2, 194, 2, 78, 228, 5, 6, 2, 2, 2, 134, 26, 4, 715, 8, 118, 2, 14, 394, 20, 13, 119, 954, 

In [82]:
# 텍스트데이터 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad=pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad=pad_sequences(X_test,maxlen=80, padding='post', truncating='post')

# RNN
from tensorflow.keras import Sequential, layers

model_rnn3=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=1000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    layers.SimpleRNN(64),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])


In [83]:
model_rnn3.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn3.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn3.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 23ms/step - accuracy: 0.5367 - loss: 0.6872
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 22ms/step - accuracy: 0.7431 - loss: 0.5306
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 21ms/step - accuracy: 0.7794 - loss: 0.4742
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step - accuracy: 0.8186 - loss: 0.4132
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.8396 - loss: 0.3744
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.8637 - loss: 0.3268
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.8818 - loss: 0.2891
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9042 - loss: 0.2400
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.9199 - loss: 0.2041
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.9388 - loss: 0.1646
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7010 - loss: 0.8875


[0.8875464797019958, 0.7009599804878235]

### 3. 영화평의 길이 200개로 변경

In [84]:
from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test)=imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [85]:
# 텍스트데이터 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad=pad_sequences(X_train, maxlen=200, padding='post', truncating='post')
X_test_pad=pad_sequences(X_test,maxlen=200, padding='post', truncating='post')

# RNN
from tensorflow.keras import Sequential, layers

model_rnn3=Sequential([
    layers.Input(shape=(200,)),
    layers.Embedding(input_dim=10000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    layers.SimpleRNN(64),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])

In [86]:
model_rnn3.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn3.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn3.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 7s 37ms/step - accuracy: 0.5025 - loss: 0.6950
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.5950 - loss: 0.6595
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.6819 - loss: 0.5284
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.7481 - loss: 0.4003
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.7684 - loss: 0.3617
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.7814 - loss: 0.3416
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 3s 23ms/step - accuracy: 0.8034 - loss: 0.3153
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 33ms/step - accuracy: 0.8243 - loss: 0.2963
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.8370 - loss: 0.2869
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 34ms/step - accuracy: 0.8503 - loss: 0.2664
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.5004 - loss: 1.3181


[1.3181178569793701, 0.5004400014877319]

### pad_sequence의 truncating과 padding을 pre로 바꾸기.

In [88]:
# 텍스트데이터 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad=pad_sequences(X_train, maxlen=80, padding='pre', truncating='pre')
X_test_pad=pad_sequences(X_test,maxlen=80, padding='pre', truncating='pre')

# RNN
from tensorflow.keras import Sequential, layers

model_rnn3=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=10000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
    layers.SimpleRNN(64),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])

In [89]:
model_rnn3.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn3.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn3.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 23ms/step - accuracy: 0.7082 - loss: 0.5384
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step - accuracy: 0.8680 - loss: 0.3134 
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9203 - loss: 0.2098
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9662 - loss: 0.1046
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9879 - loss: 0.0451
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9971 - loss: 0.0156
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9994 - loss: 0.0055
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 18ms/step - accuracy: 0.9996 - loss: 0.0036
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 1.0000 - loss: 0.0011
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 1.0000 - loss: 4.2191e-04
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8120 - loss: 0.9640


[0.9640120267868042, 0.8119999766349792]

### 5. RNN 층(뉴런 128개)을 하나 더 추가해

In [90]:
# 텍스트데이터 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad=pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad=pad_sequences(X_test,maxlen=80, padding='post', truncating='post')

from tensorflow.keras import Sequential, layers
model_rnn3=Sequential([
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=1000, output_dim=32), #만개의 단어, 각 단어를 32차원 벡터로 매핑
     layers.SimpleRNN(64, return_sequences=True),
    layers.SimpleRNN(128),
    layers.Dense(2, activation='softmax') # 이진분류(클래스 2개)
])

In [91]:
model_rnn3.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_rnn3.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_rnn3.evaluate(X_test_pad, y_test)

Epoch 1/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 7s 33ms/step - accuracy: 0.6683 - loss: 0.5966
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.7745 - loss: 0.4806
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.7870 - loss: 0.4557
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 2s 19ms/step - accuracy: 0.7986 - loss: 0.4442
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.8114 - loss: 0.4177
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.8352 - loss: 0.3740
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.8547 - loss: 0.3412
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 0.8721 - loss: 0.3020
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - accuracy: 0.8936 - loss: 0.2610
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - accuracy: 0.9120 - loss: 0.2188
782/782 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.7423 - loss: 0.7651


[0.7650826573371887, 0.7422800064086914]

| 실험 번호 | Loss 값  | Accuracy 값 | 설명 |
|-----------|----------|-------------|------|
| 1번       | 0.6930   | 0.508 (50.8%) | 옵티마이저를 **SGD**로 사용 → 수렴이 잘 안 되어 정확도가 낮음 (무작위 예측 수준). |
| 2번       | 0.8875   | 0.701 (70.1%) | **단어 개수를 1000개로 제한** → 희소 단어 제거 효과, 성능이 개선됨. |
| 3번       | 1.3181   | 0.500 (50.0%) | **리뷰 길이를 200으로 증가** → 불필요한 단어·패딩 증가, RNN의 기울기 소실 문제 심화 → 성능 급락. |
| 4번       | 0.9640   | 0.812 (81.2%) | **padding/truncating='pre'** 적용 → 문장의 뒷부분(핵심 내용) 보존 효과 → 성능 크게 향상. |
| 5번       | 0.7651   | 0.742 (74.2%) | **RNN 층 추가 (128 유닛)** → 표현력은 증가했지만 파라미터 수 증가로 과적합 위험 → 성능이 4번보다 낮음. |


## CNN

In [70]:

from tensorflow.keras.datasets import imdb
(X_train, y_train), (X_test, y_test)=imdb.load_data(num_words=10000)
(X_train, y_train), (X_test, y_test)

((array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 194, 1153, 194, 8255, 78, 

In [71]:
# 텍스트데이터 전처리
from tensorflow.keras.preprocessing.sequence import pad_sequences
X_train_pad=pad_sequences(X_train, maxlen=80, padding='post', truncating='post')
X_test_pad=pad_sequences(X_test,maxlen=80, padding='post', truncating='post')

In [74]:
from tensorflow.keras import Sequential, layers

model_cnn=Sequential(
[
    layers.Input(shape=(80,)),
    layers.Embedding(input_dim=10000, output_dim=32),
    layers.Conv1D(64,3, activation='relu'),
    layers.GlobalMaxPool1D(),
    layers.Dense(64, activation='relu'),
    layers.Dense(2, activation='softmax'),
])

In [75]:
model_cnn.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_cnn.fit(X_train_pad, y_train, epochs=10, batch_size=200)
model_cnn.evaluate(X_test_pad, y_test)

Epoch 1/10


2025-09-04 16:34:43.028956: I external/local_xla/xla/service/gpu/autotuning/dot_search_space.cc:208] All configs were filtered out because none of them sufficiently match the hints. Maybe the hints set does not contain a good representative set of valid configs? Working around this by using the full hints set instead.
2025-09-04 16:34:43.029014: I external/local_xla/xla/service/gpu/autotuning/dot_search_space.cc:208] All configs were filtered out because none of them sufficiently match the hints. Maybe the hints set does not contain a good representative set of valid configs? Working around this by using the full hints set instead.
2025-09-04 16:34:43.664960: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_389', 8 bytes spill stores, 8 bytes spill loads

2025-09-04 16:34:43.765855: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Re

125/125 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.6670 - loss: 0.6096
Epoch 2/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8286 - loss: 0.3891
Epoch 3/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9030 - loss: 0.2494
Epoch 4/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9556 - loss: 0.1366
Epoch 5/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9860 - loss: 0.0579
Epoch 6/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9984 - loss: 0.0193
Epoch 7/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9997 - loss: 0.0069
Epoch 8/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 1.0000 - loss: 0.0031
Epoch 9/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 1.0000 - loss: 0.0019
Epoch 10/10
125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 1.0000 - loss: 0.0013
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7860 - loss: 0.9444


[0.9444447159767151, 0.7860000133514404]